### PR TITLE
Add interfaces and improve block generation code

### PR DESCRIPTION
**Pull request by ben-chain**
## Description
This PR adds to the ``MerkleIntervalTree`` proper interfaces for inclusion proofs, tree nodes, etc.  It also improves on the branch verification logic so that the inclusion proof checking for the double interval tree is done in a more logical way.  Also, there's accidentally some preliminary commitment contract code I forgot I had committed, so please ignore 😅

## Questions
- One naming question: I called the interface ``DoubleMerkleIntervalTree`` and the class ``PlasmaBlock``.  Do we prefer something more like ``PlasmaBlock`` and ``GenericPlasmaBlock`` instead?

## Contributing Agreement

- [x] I have read and understood the [Plasma Group Contributing Guide and Code of Conduct](https://github.com/plasma-group/pigi/blob/master/.github/CONTRIBUTING.md) and am following those guidelines in this pull request.

### CODE REVIEW

File: packages/core/test/app/block-production/merkle-interval-tree.spec.ts [hunk 1]
```diff
@@ -48,97 +69,97 @@ describe('merkle-index-tree', () => {
...
-          new BigNum(1),
-          '0xbdAd2846585129Fc98538ce21cfcED21dDDE0a63'
-        )
-        stateUpdates.push(stateUpdate)
-      }
+      const stateUpdates = generateNSequentialStateUpdates(4)
       const merkleStateIntervalTree = new MerkleStateIntervalTree(stateUpdates)
       log('root', merkleStateIntervalTree.root())
     })
+    it('verification should throw if state update range intersects branch bounds', async () => {
```
Comment (karlfloersch) [inline packages/core/test/app/block-production/merkle-interval-tree.spec.ts]: Nitpick but I'd change the test name from  "it(verification should throw if..." -> to
```
describe('verifyExectedRoot') {
  it('should throw if state update range...')
}
```

A little more grammatically correct

File: packages/core/src/types/block-production.types.ts [hunk 2]
```diff
@@ -4,7 +4,44 @@ import BigNumber = require('bn.js')
...
+  root: MerkleIntervalTreeNode
+  maxEnd: BigNumber
+}
+
+export interface MerkleIntervalTree {
+  dataBlocks: any
+  levels: MerkleIntervalTreeNode[][]
+  root(): MerkleIntervalTreeNode
+  getInclusionProof(leafposition: number): MerkleIntervalInclusionProof
+}
```
Comment (karlfloersch) [inline packages/core/src/types/block-production.types.ts]: Would be good to add comments to these variables -- especially `levels` and `maxEnd` as those are pretty specific to our designs

File: packages/core/src/types/block-production.types.ts [hunk 3]
```diff
@@ -4,7 +4,44 @@ import BigNumber = require('bn.js')
...
+export interface DoubleMerkleIntervalTree extends MerkleIntervalTree {
+  dataBlocks: SubtreeContents
+  subtrees: MerkleIntervalTree[]
+  getStateUpdateInclusionProof(
+    stateUpdatePosition: number,
+    assetIdPosition: number
+  ): DoubleMerkleInclusionProof
+}
+
+export interface DoubleMerkleInclusionProof {
```
Comment (karlfloersch) [inline packages/core/src/types/block-production.types.ts]: Would be good to add a comment with a link to our spec http://spec.plasma.group/en/latest/src/01-core/double-layer-tree.html

File: packages/contracts/test/Commitment.spec.ts [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+      //   )
+      //   //   console.log('typescript left sibling is ', tsLeftSibling)
+      //   const tsRightSibling = new MerkleIntervalTreeNode(
+      //     Buffer.from(rightSibling.hashValue.slice(2), 'hex'),
+      //     new BigNum(rightSibling.start.slice(2), 'hex').toBuffer('be', 16)
+      //   )
+      //   //   console.log('typescript right sibling is ', tsRightSibling)
+
+      //   const tsParent = MerkleIntervalTree.parent(tsLeftSibling, tsRightSibling)
+      //   //   console.log('typescript parent is: ', tsParent)
```
[the remark this answers is not in the document]
Comment (ben-chain) [reply packages/contracts/test/Commitment.spec.ts]: Just removed this file entirely since unused in this PR.

**Comment by ben-chain** [comment]
Made the requested changes, should be good to go!

File: packages/core/src/types/block-production.types.ts [hunk 5]
```diff
@@ -4,7 +4,48 @@ import BigNumber = require('bn.js')
...
+export interface MerkleIntervalProofOutput {
+  root: MerkleIntervalTreeNode // the root node resulting from a merkle index tree inclusion proof
+  upperBound: BigNumber // The upper bound that an inclusion proof is valid for.
+  // For a single MerkleIntervalTree, it is mathematically impossible for two branches to exist
+  // such that their [leaf.lowerBound, proofOutput.upperBound) intersect.
+}
+
+export interface MerkleIntervalTree {
+  dataBlocks: any // The blocks of data we are constructing a merkle interval tree for.
+  levels: MerkleIntervalTreeNode[][] // The 'MerkleIntervalTreeNode's which make up the tree. levels[0].length ==
```
Comment (karlfloersch) [inline packages/core/src/types/block-production.types.ts]: > levels[0].length ==

Looks like a typo

**Pull request merged**